### PR TITLE
[BUGFIX] Raise an error for no draft token case when draft_tp>1

### PR DESCRIPTION
**Pull request by wooyeonlee0**
This PR adds a simple patch to raise an error to prevent users from hitting the hang error stated in #5814.
This error happens only when the skip speculation feature is activated and there are no generated draft tokens for "all" sequences in a step, and draft_tp > 1.

We may revisit this issue later because it's not resolved completely.

### CODE REVIEW

**Comment by wooyeonlee0** [comment]
@cadedaniel @zifeitong @comaniac
This PR is to implement the second option (raise an error) in cade's suggestion.
Could any one of you review this?
Maybe I can revisit later to completely fix it, following the first option.

> We need to either fix it or raise an error so that users don't hit this.
https://github.com/vllm-project/vllm/pull/5414#issuecomment-2187670183

**Review by comaniac — COMMENTED** [review]
LGTM. Can we enable the test and expect to catch this exception?

**Comment by wooyeonlee0** [comment]
@comaniac Thanks for the review :)
I've added the test code that catches the error.

But I'm not sure, the CI seems to have stopped.

**Comment by wooyeonlee0** [comment]
@comaniac The CI test passed :)

File: vllm/spec_decode/spec_decode_worker.py [hunk 1]
```diff
@@ -461,6 +471,12 @@ def _run_speculative_decoding_step(
         proposals = self.proposer_worker.get_spec_proposals(
             execute_model_req, self._seq_with_bonus_token_in_last_step)
 
+        if not self.allow_no_draft_tokens and sum(
+                proposals.proposal_lens) == 0:
```
Comment (comaniac) [inline vllm/spec_decode/spec_decode_worker.py:475]: I'm a bit worry about the overhead that `sum` brings, but I feel it's fine for now given that it won't be triggered with draft model TP=1. wdyt @cadedaniel?

**Review by comaniac — APPROVED** [review]
LGTM

File: vllm/spec_decode/spec_decode_worker.py [hunk 2]
```diff
@@ -187,11 +193,15 @@ def __init__(
...
                 for testing purposes.
+            allow_zero_draft_token_step: whether to allow a step where the draft
+                model generates no draft token; should disallow when the tp of
+                draft model is larger than 1 (TODO: #5814)
         """
         self.proposer_worker = proposer_worker
         self.scorer_worker = scorer_worker
         self.disable_by_batch_size = disable_by_batch_size or float("inf")
         self.spec_decode_sampler = spec_decode_sampler
+        self.allow_no_draft_tokens = allow_zero_draft_token_step
```
Comment (cadedaniel) [inline vllm/spec_decode/spec_decode_worker.py:204]: nit: can we mark this private, e.g. `_allow_no_draft_tokens`? we should have done this for all properties but we missed it

File: vllm/spec_decode/spec_decode_worker.py (hunk 1 above)
Comment (cadedaniel) [reply vllm/spec_decode/spec_decode_worker.py:475]: yeah, is it possible to store this field in `proposals` when we create `proposals`? that way we don't need an additional CPU-GPU-CPU sync

**Comment by wooyeonlee0** [comment]
Thanks for the review!  I'll handle it right now :)

**Comment by wooyeonlee0** [comment]
@comaniac @cadedaniel
I've updated the code so that the proposer returns None when there's no proposal.
I'll let you know when the CI succeeds!

**Comment by wooyeonlee0** [comment]
@comaniac 
I re-initiated the CI test multiple times, but CI failed in one of the following cases: 'build image' or 'documentation build'.
Link: https://buildkite.com/vllm/ci-aws/builds/5151#0190c4c4-6cde-4a9f-a587-63f29a3b5dbd
Link: https://buildkite.com/vllm/ci-aws/builds/5218#0190c8b9-3ee9-4548-a691-cbe56abcff24

Is there any problem in CI now?

**Comment by comaniac** [comment]
The failure you posted seems random. I'll monitor the current CI run and manually retry failed ones.

**Comment by wooyeonlee0** [comment]
@cadedaniel @comaniac I've updated the code as your suggestion and I think the code passed the test.
Would you take a look? :)

@comaniac CI has finished. Would you retry the 'documentation-build' test? Thank you!

**Review by cadedaniel — APPROVED** [review]

**Comment by cadedaniel** [comment]
@simon-mo can we get a force merge, doc build seems broken

**Pull request merged**